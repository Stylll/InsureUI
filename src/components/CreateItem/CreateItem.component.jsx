import React, { useState, useEffect, useRef } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import * as itemActions from '../../redux/actionCreators/itemActions/itemActions';
import TextInput from '../TextInput/TextInput.component';
import Button from '../Button/Button.component';
import Dropdown from '../Dropdown/Dropdown.component';
import { validateCreateItem } from '../../helpers/validators/itemValidator';
import { categories } from '../../helpers/defaults';
import './CreateItem.styles.scss';

const CreateItem = ({
  createItem, isLoading: reduxIsLoading, closeModal,
  errorMessage, createItemSuccess, dataTestId,
}) => {
  const [values, setValues] = useState({
    Name: '',
    Value: '',
    CategoryId: '',
  });

  const [formErrors, setFormErrors] = useState({
    Name: '',
    Value: '',
    CategoryId: '',
  });

  const [isLoading, setIsLoading] = useState(false);

  const [serverError, setServerError] = useState('');

  const formSubmitted = useRef(false);
  const hasServerOperationOccurred = useRef(false);

  useEffect(() => {
    if (!reduxIsLoading && createItemSuccess && formSubmitted.current) {
      setIsLoading(false);
      closeModal();
      formSubmitted.current = false;
      return;
    }

    if (!reduxIsLoading && formSubmitted.current) {
      setIsLoading(false);
      formSubmitted.current = false;
    }
  }, [reduxIsLoading, createItemSuccess, closeModal]);

  useEffect(() => {
    if (errorMessage && hasServerOperationOccurred.current) {
      setServerError(errorMessage);
    }
  }, [errorMessage]);

  const handleChange = (event) => {
    event.preventDefault();
    setValues({
      ...values,
      [event.target.name]: event.target.value,
    });
  };

  const handleClick = (event) => {
    event.preventDefault();
    const { isValid, errors } = validateCreateItem(values);
    setFormErrors({
      ...errors,
    });
    if (!isValid) {
      return;
    }

    setIsLoading(true);

    createItem({
      Name: values.Name,
      Value: values.Value,
      CategoryId: values.CategoryId,
    });

    formSubmitted.current = true;
    hasServerOperationOccurred.current = true;
  };

  const renderCreateForm = () => (
    <div className="item-container">
        <div>
            <p>Add an Item</p>
            <span className="error-text">{serverError}</span>
            <TextInput
                name="Name"
                placeholder="Playstation"
                title="Enter the item name"
                type="text"
                value={values.Name}
                error={formErrors.Name}
                onChange={handleChange}
                dataTestId={`${dataTestId}-name`} />
            <TextInput
                name="Value"
                placeholder="5000"
                title="Enter the value of the item"
                type="text"
                value={values.Value}
                error={formErrors.Value}
                onChange={handleChange}
                dataTestId={`${dataTestId}-value`} />
            <Dropdown
                options={categories}
                name="CategoryId"
                optionsTitleProperty="name"
                optionsValueProperty="id"
                value={values.CategoryId}
                error={formErrors.CategoryId}
                onChange={handleChange}
                placeholder="select item category"
                dataTestId={`${dataTestId}-category`}
            />
            <br />
            <Button
                title="Create"
                handleClick={handleClick}
                showLoader={isLoading}
                disabled={isLoading}
                dataTestId={`${dataTestId}-create`}
            />
        </div>
        <br />
    </div>
  );

  return (
        <div className="create-item-container">
            <div className="form-container">
                    {renderCreateForm()}
                </div>
        </div>
  );
};

CreateItem.propTypes = {
  createItem: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  createItemSuccess: PropTypes.bool.isRequired,
  errorMessage: PropTypes.string.isRequired,
  dataTestId: PropTypes.string,
};

CreateItem.defaultProps = {
  dataTestId: 'createitem',
};

const mapStateToProps = ({ item }) => ({
  isLoading: item.isLoading,
  createItemSuccess: item.createItemSuccess,
  errorMessage: item.errorMessage,
});

const mapDispatchToProps = {
  createItem: itemActions.createItem,
};

export {
  CreateItem,
};

export default connect(mapStateToProps, mapDispatchToProps)(CreateItem);
