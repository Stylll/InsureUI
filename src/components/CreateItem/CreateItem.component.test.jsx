import { render, fireEvent } from '@testing-library/react';
import React from 'react';

import { CreateItem } from '../../components/CreateItem/CreateItem.component';

describe('CreateItem Test', () => {
  it('should render properly', () => {
    const { baseElement } = render(<CreateItem
      createItem={() => {}}
      closeModal={() => {}}
      isLoading={false}
      createItemSuccess={false}
      errorMessage=""
    />);
    expect(baseElement).toMatchSnapshot();
  });
  it('should validate the input', () => {
    const { baseElement, getByTestId, getByText } = render(<CreateItem
      createItem={() => {}}
      closeModal={() => {}}
      isLoading={false}
      createItemSuccess={false}
      errorMessage=""
    />);
    fireEvent.click(getByTestId('createitem-create'));
    expect(getByText('name cannot be empty')).toBeInTheDocument();
    expect(getByText('value cannot be empty')).toBeInTheDocument();
    expect(getByText('category must be either Electronics, Clothing, or Kitchen')).toBeInTheDocument();
  });

  it('should submit form, call createItem and display error', () => {
    const createItem = jest.fn();
    const {
      baseElement, getByTestId, getByText, rerender,
    } = render(<CreateItem
      createItem={createItem}
      closeModal={() => {}}
      isLoading={false}
      createItemSuccess={false}
      errorMessage=""
    />);

    fireEvent.change(getByTestId('createitem-name'),
      { target: { value: 'Playstation', name: 'Name' } });
    fireEvent.change(getByTestId('createitem-value'),
      { target: { value: '245', name: 'Value' } });
    fireEvent.change(getByTestId('createitem-category'),
      { target: { value: '1', name: 'CategoryId' } });
    fireEvent.click(getByTestId('createitem-create'));
    expect(createItem).toHaveBeenCalled();

    rerender(<CreateItem
      createItem={createItem}
      closeModal={() => {}}
      isLoading={false}
      createItemSuccess={false}
      errorMessage="An error occurred"
    />);
    expect(getByText('An error occurred')).toBeInTheDocument();
  });
});
