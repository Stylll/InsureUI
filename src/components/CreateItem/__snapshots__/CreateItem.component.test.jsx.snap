// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateItem Test should render properly 1`] = `
<body>
  <div>
    <div
      class="create-item-container"
    >
      <div
        class="form-container"
      >
        <div
          class="item-container"
        >
          <div>
            <p>
              Add an Item
            </p>
            <span
              class="error-text"
            />
            <div
              class="form-input"
            >
              <div
                class="title"
              >
                Enter the item name
              </div>
              <input
                class="text-input"
                data-testid="createitem-name"
                id=""
                name="Name"
                placeholder="Playstation"
                type="text"
                value=""
              />
              <span
                class="info"
              />
              
            </div>
            <div
              class="form-input"
            >
              <div
                class="title"
              >
                Enter the value of the item
              </div>
              <input
                class="text-input"
                data-testid="createitem-value"
                id=""
                name="Value"
                placeholder="5000"
                type="text"
                value=""
              />
              <span
                class="info"
              />
              
            </div>
            <div
              class="drop-down"
            >
              <select
                class="select"
                data-testid="createitem-category"
                name="CategoryId"
              >
                <option
                  value=""
                >
                  select item category
                </option>
                <option
                  value="1"
                >
                  Electronics
                </option>
                <option
                  value="2"
                >
                  Clothing
                </option>
                <option
                  value="3"
                >
                  Kitchen
                </option>
              </select>
              <span
                class="chevy"
              >
                âˆ¨
              </span>
              <span
                class="info"
              />
              
            </div>
            <br />
            <div
              class="btn-button"
            >
              <button
                class="button "
                data-testid="createitem-create"
                type="submit"
              >
                Create
              </button>
            </div>
          </div>
          <br />
        </div>
      </div>
    </div>
  </div>
</body>
`;
